### PR TITLE
ARROW-4264: [C++] Clarify use of DCHECKs in Kernels

### PR DESCRIPTION
**Pull request by emkornfield**
Also make format caught some random changes.

@wesm @fsaintjacques I'm about 90% agreed this approach is reasonable but I'm not sure this is very important to work out now (it should be pretty quick to reverse this decision), but we should document so we can at least be consistent in implementing kernels.

### CODE REVIEW

**Review by fsaintjacques — APPROVED** [review]

**Comment by xhochy** [comment]
Rebased.

File: cpp/src/arrow/compute/kernel.h [hunk 1]
```diff
@@ -36,6 +36,20 @@ class FunctionContext;
...
+///
+/// Note to implementors:
+/// Operator kernels are intended to be the lowest level of an analytics/compute
+/// engine.  They will generally not be exposed directly to end-users.  Instead
+/// they will be wrapped by higher level constructs (e.g. function or physical
+/// execution plan nodes).  These higher level constructs are
+/// responsible for user input validation and returning the appropriate
+/// error Status.
+///
+/// Due to this design, implementations of of Call (the execution
```
Comment (aregm) [inline cpp/src/arrow/compute/kernel.h]: double "of"

**Comment by wesm** [comment]
@emkornfield I think you formatted with clang-format-6.0. I rebased and ran clang-format-7

**Comment by wesm** [comment]
+1. there was a case of ARROW-3344 that failed the manylinux1 build

**Pull request closed without merging**